File: packages/typescript-sdk/src/transport.ts
import type { z } from "zod";
import {
  StarciteApiError,
  StarciteConnectionError,
  StarciteError,
} from "./errors";
import type { SocketManager } from "./socket-manager";

const TRAILING_SLASHES_RE = /\/+$/;

/**
 * Shared HTTP + WebSocket transport configuration.
 *
 * Both `Starcite` (API-key auth) and `StarciteSession` (session-token auth)
 * hold their own `TransportConfig` and pass it to the free functions below.
 */
export interface TransportConfig {
  readonly baseUrl: string;
  readonly socketManager: SocketManager;
  bearerToken: string | null;
  readonly fetchFn: typeof fetch;
}

/**
 * Validates a URL has an http/https protocol, strips trailing slashes, and returns the URL object.
 */
export function parseHttpUrl(value: string): URL {
  const url = new URL(value);

  if (url.protocol !== "http:" && url.protocol !== "https:") {
    throw new StarciteError(`URL must use http:// or https://: ${value}`);
  }

  url.pathname = url.pathname.replace(TRAILING_SLASHES_RE, "");
  return url;
}

/**
 * Converts a Starcite base URL to the `/v1` API root used by this SDK.
 */
export function toApiBaseUrl(baseUrl: string): string {
  const value = stripTrailingSlashes(parseHttpUrl(baseUrl).toString());
  return value.endsWith("/v1") ? value : `${value}/v1`;
}

/**
 * Converts HTTP API base URL to its websocket equivalent.
 */
export function toWebSocketBaseUrl(apiBaseUrl: string): string {
  const url = parseHttpUrl(apiBaseUrl);
  url.protocol = url.protocol === "https:" ? "wss:" : "ws:";
  return stripTrailingSlashes(url.toString());
}

export function stripTrailingSlashes(value: string): string {
  return value.replace(TRAILING_SLASHES_RE, "");
}

/**
 * Makes an HTTP request against the transport's base URL.
 */
export function request<T>(
  transport: TransportConfig,
  path: string,
  init: RequestInit,
  schema: z.ZodType<T>
): Promise<T> {
  return requestWithBaseUrl(transport, transport.baseUrl, path, init, schema);
}

/**
 * Makes an HTTP request against an arbitrary base URL using the transport's shared config.
 */
export async function requestWithBaseUrl<T>(
  transport: TransportConfig,
  baseUrl: string,
  path: string,
  init: RequestInit,
  schema: z.ZodType<T>
): Promise<T> {
  const headers = new Headers();

  if (transport.bearerToken) {
    headers.set("authorization", `Bearer ${transport.bearerToken}`);
  }

  if (init.body !== undefined && !headers.has("content-type")) {
    headers.set("content-type", "application/json");
  }

  if (init.headers) {
    const perRequestHeaders = new Headers(init.headers);
    for (const [key, value] of perRequestHeaders.entries()) {
      headers.set(key, value);
    }
  }

  let response: Response;

  try {
    response = await transport.fetchFn(`${baseUrl}${path}`, {
      ...init,
      headers,
    });
  } catch (error) {
    throw new StarciteConnectionError(
      `Failed to connect to Starcite at ${baseUrl}: ${error instanceof Error ? error.message : String(error)}`
    );
  }

  if (!response.ok) {
    let payload: Record<string, unknown> | null = null;
    try {
      payload = (await response.json()) as Record<string, unknown>;
    } catch {
      // Non-JSON error response — fall through with null payload.
    }

    const code =
      typeof payload?.error === "string"
        ? payload.error
        : `http_${response.status}`;
    const message =
      typeof payload?.message === "string"
        ? payload.message
        : response.statusText;

    throw new StarciteApiError(message, response.status, code, payload);
  }

  if (response.status === 204) {
    return schema.parse(undefined);
  }

  let body: unknown;
  try {
    body = await response.json();
  } catch (error) {
    throw new StarciteConnectionError(
      `Received invalid JSON from Starcite: ${error instanceof Error ? error.message : String(error)}`
    );
  }

  return schema.parse(body);
}

File: packages/python-sdk/src/starcite_sdk/transport.py
from __future__ import annotations

import json
from collections.abc import Awaitable, Callable
from dataclasses import dataclass
from typing import TypeVar
from urllib.parse import urljoin, urlparse

import httpx

from .errors import StarciteApiError, StarciteConnectionError, StarciteError
from .types import RequestOptions

T = TypeVar("T")
Parser = Callable[[object], T]


@dataclass(frozen=True)
class HttpRequest:
    method: str
    url: str
    headers: dict[str, str]
    body: bytes | None = None
    timeout_seconds: float | None = None


@dataclass(frozen=True)
class HttpResponse:
    status_code: int
    body: bytes
    headers: dict[str, str]
    reason: str = ""


AsyncRequester = Callable[[HttpRequest], Awaitable[HttpResponse]]


@dataclass(frozen=True)
class TransportConfig:
    base_url: str
    websocket_url: str
    requester: AsyncRequester
    bearer_token: str | None = None


def strip_trailing_slashes(value: str) -> str:
    return value.rstrip("/")


def parse_http_url(value: str) -> str:
    parsed = urlparse(value)
    if parsed.scheme not in {"http", "https"}:
        raise StarciteError(f"URL must use http:// or https://: {value}")
    if parsed.netloc == "":
        raise StarciteError(f"URL must include a host: {value}")
    path = strip_trailing_slashes(parsed.path)
    return parsed._replace(path=path).geturl()


def to_api_base_url(base_url: str) -> str:
    normalized = strip_trailing_slashes(parse_http_url(base_url))
    return normalized if normalized.endswith("/v1") else f"{normalized}/v1"


def to_websocket_base_url(api_base_url: str) -> str:
    parsed = urlparse(parse_http_url(api_base_url))
    scheme = "wss" if parsed.scheme == "https" else "ws"
    return strip_trailing_slashes(parsed._replace(scheme=scheme).geturl())


def to_phoenix_websocket_url(api_base_url: str) -> str:
    return f"{to_websocket_base_url(api_base_url)}/socket/websocket"


def make_httpx_requester(client: httpx.AsyncClient) -> AsyncRequester:
    async def requester(request: HttpRequest) -> HttpResponse:
        response = await client.request(
            method=request.method,
            url=request.url,
            headers=request.headers,
            content=request.body,
            timeout=request.timeout_seconds,
        )
        return HttpResponse(
            status_code=response.status_code,
            body=response.content,
            headers=dict(response.headers.items()),
            reason=response.reason_phrase,
        )

    return requester


async def request_with_base_url(
    transport: TransportConfig,
    *,
    base_url: str,
    path: str,
    method: str,
    parser: Parser[T],
    body: dict[str, object] | None = None,
    headers: dict[str, str] | None = None,
    request_options: RequestOptions | None = None,
) -> T:
    request_headers: dict[str, str] = {}
    if transport.bearer_token:
        request_headers["authorization"] = f"Bearer {transport.bearer_token}"
    if body is not None:
        request_headers["content-type"] = "application/json"
    if headers:
        request_headers.update(headers)

    request_body = json.dumps(body).encode("utf-8") if body is not None else None
    url = urljoin(f"{strip_trailing_slashes(base_url)}/", path.lstrip("/"))

    try:
        response = await transport.requester(
            HttpRequest(
                method=method,
                url=url,
                headers=request_headers,
                body=request_body,
                timeout_seconds=request_options.timeout_seconds if request_options else None,
            )
        )
    except StarciteError:
        raise
    except httpx.HTTPError as exc:
        raise StarciteConnectionError(
            f"Failed to connect to Starcite at {base_url}: {exc}"
        ) from exc
    except Exception as exc:
        raise StarciteConnectionError(
            f"Failed to connect to Starcite at {base_url}: {exc}"
        ) from exc

    payload: dict[str, object] | None = None
    if response.body:
        try:
            decoded = json.loads(response.body.decode("utf-8"))
            if isinstance(decoded, dict):
                payload = decoded
        except Exception as exc:
            if 200 <= response.status_code < 300:
                raise StarciteConnectionError(
                    f"Received invalid JSON from Starcite: {exc}"
                ) from exc

    if not 200 <= response.status_code < 300:
        code = (
            payload.get("error")
            if isinstance(payload, dict) and isinstance(payload.get("error"), str)
            else f"http_{response.status_code}"
        )
        message = (
            payload.get("message")
            if isinstance(payload, dict) and isinstance(payload.get("message"), str)
            else response.reason
            or f"HTTP {response.status_code}"
        )
        raise StarciteApiError(message, response.status_code, code, payload)

    if response.status_code == 204:
        return parser(None)

    try:
        return parser(payload)
    except (TypeError, ValueError) as exc:
        raise StarciteConnectionError(
            f"Received invalid JSON from Starcite: {exc}"
        ) from exc


async def request_json(
    transport: TransportConfig,
    *,
    path: str,
    method: str,
    parser: Parser[T],
    body: dict[str, object] | None = None,
    headers: dict[str, str] | None = None,
    request_options: RequestOptions | None = None,
) -> T:
    return await request_with_base_url(
        transport,
        base_url=transport.base_url,
        path=path,
        method=method,
        parser=parser,
        body=body,
        headers=headers,
        request_options=request_options,
    )
